File: new/15_3sum_two_pointers.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        // idea: two pointer
        // get the firs value (should not be positive, because in a sorted list of numbers, 
        // if the first num is +ve, the second and third will also be +ve, and
        // no sum of 3 positive numbers will be 0). 
        // iterate after the first value and get two more values (second and third num)
        // using two pointer method. If the sum adds up to 0, we have a valid triplet.
        sort(nums.begin(), nums.end());
        vector<vector<int>> valid_triplets;

        for (int i = 0; i < nums.size(); i++) {
            if (nums[i] > 0) break;
            if (i > 0 && nums[i] == nums[i - 1]) continue;
            int left = i + 1;
            int right = nums.size() - 1;
            while (left < right) {
                int total = nums[i] + nums[left] + nums[right];
                if (total > 0) {
                    right -= 1;
                } else if (total < 0) {
                    left += 1;
                } else {
                    valid_triplets.push_back({nums[i], nums[left], nums[right]});
                    left++;
                    right--;
                    while (left < right && nums[left] == nums[left - 1]) {
                        left++;
                    }
                    while (right > left && nums[right] == nums[right + 1]) {
                        right--;
                    }
                }
            }
        }

        return valid_triplets;
    }
    // time: O(n^2)
    // space: O(1 or n) depending on the sorting algorithm, O(m) for extra space for the output
    // where m is the length of number of triplets that are valid. 
};

File: new/python/15_3sum_two_pointers.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        # idea: two pointers
        # for every first number (should not be positive) find two other number using
        # two pointers for which the the three numbers together sum up to 0. After sorting
        # if the first number is not 0 or -ve value, then the following two numbers will also be
        # positive. No sum of three positive numbers will be positive.
        nums.sort()
        valid_triplet = []
        for i, val in enumerate(nums):
            if val > 0:  # if first number of the triplet from sorted array is +ve, we
                # cannot have sum of three number to be 0.
                break
            if i > 0 and nums[i] == nums[i - 1]:
                continue

            left = i + 1
            right = len(nums) - 1
            while left < right:
                total = val + nums[left] + nums[right]
                if total > 0:
                    right -= 1
                elif total < 0:
                    left += 1
                else:
                    valid_triplet.append([val, nums[left], nums[right]])
                    left += 1
                    right -= 1
                    while left < right and nums[left] == nums[left - 1]:
                        left += 1
                    while right > left and nums[right] == nums[right + 1]:
                        right -= 1

        return valid_triplet
        # time: O(n^2)
        # space: O(1 or n) depending or sorting algorithm nums.sort()
        #        O(m) for extra space for the output list
